File: packages/nakdimon/src/nakdimon/runtime.py
"""ONNX inference wrapper. Mirrors the v1 predict pipeline (nakdimon/predict.py +
nakdimon/dataset.py) exactly, so that byte-exact reproduction of the shipped v1
model's outputs (spec/golden/predict_v1.json) is a build invariant, not an accident.
"""

from __future__ import annotations

import json
from collections.abc import Iterator
from dataclasses import dataclass, fields
from pathlib import Path

import numpy as np
import onnxruntime as ort

from nakdimon.encode import Inventory
from nakdimon.hebrew import RAFE, HebrewItem, can_dagesh, can_niqqud, can_sin, clean_text, decompose
from nakdimon.spec import load_spec


@dataclass(frozen=True)
class ModelCard:
    inventory: str = "v1"
    window: int = 10000
    pad_to_window: bool = True
    overlap: int = 0

    @staticmethod
    def load(model_path: str | Path) -> ModelCard:
        sidecar = Path(f"{model_path}.json")
        if not sidecar.exists():
            return ModelCard()
        data = json.loads(sidecar.read_text(encoding="utf-8"))
        # Sidecars may carry extra provenance fields (config, sha256, train_run, ...);
        # only the windowing/inventory contract is this runtime's concern.
        known = {f.name for f in fields(ModelCard)}
        return ModelCard(**{k: v for k, v in data.items() if k in known})


def _split_by_length(items: list[HebrewItem], maxlen: int) -> Iterator[list[HebrewItem]]:
    """Chunk items into windows of at most `maxlen`, preferring to break on the last
    space seen so words are never split across chunk boundaries. Ported verbatim from
    v1's hebrew.split_by_length: a subtly different boundary rule changes chunking and
    therefore the model's output.
    """
    assert maxlen > 1
    out: list[HebrewItem] = []
    space = maxlen
    for item in items:
        if item.letter == " ":
            space = len(out)
        out.append(item)
        if len(out) == maxlen - 1:
            yield out[: space + 1]
            out = out[space + 1 :]
    if out:
        yield out


def _window_starts(n: int, window: int, overlap: int) -> list[int]:
    """Start offsets of overlapping windows covering [0, n). The last window is
    right-aligned so no position is ever closer than `overlap`/2 to a hard edge
    (except the text's own ends)."""
    if n <= window:
        return [0]
    stride = window - overlap
    starts = list(range(0, n - window, stride))
    starts.append(n - window)
    return starts


def _ownership_cuts(starts: list[int], window: int, n: int) -> list[int]:
    """Exclusive end of each window's owned region: consecutive windows meet at
    the midpoint of their overlap, so every position is predicted by the window
    in which it is most central. Integer floor midpoint — keep identical to the
    JS port."""
    cuts = [(starts[k + 1] + starts[k] + window) // 2 for k in range(len(starts) - 1)]
    cuts.append(n)
    return cuts


class Diacritizer:
    def __init__(self, model_path: str | Path, providers: list[str] | None = None) -> None:
        session = ort.InferenceSession(str(model_path), providers=providers or ["CPUExecutionProvider"])
        self._init(session, ModelCard.load(model_path))

    @classmethod
    def from_session(cls, session, card: ModelCard) -> Diacritizer:
        """Wrap an existing (possibly fake) session — the testing/embedding hook."""
        self = cls.__new__(cls)
        self._init(session, card)
        return self

    def _init(self, session, card: ModelCard) -> None:
        self.session = session
        self.card = card
        self.inventory = Inventory(load_spec()["inventories"][card.inventory])
        self._input_name = session.get_inputs()[0].name

    def _run(self, input_array: np.ndarray) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
        n_out, d_out, s_out = self.session.run(None, {self._input_name: input_array})
        return n_out.argmax(axis=-1), d_out.argmax(axis=-1), s_out.argmax(axis=-1)

    def _emit(self, item: HebrewItem, niqqud_id: int, dagesh_id: int, sin_id: int) -> str:
        # NFC canonical emission order: niqqud, dagesh, sin-dot (spec["mark_order"]).
        parts = [item.letter]
        if can_niqqud(item.letter):
            parts.append(self.inventory.decode_niqqud(niqqud_id))
        if can_dagesh(item.letter):
            parts.append(self.inventory.decode_dagesh(dagesh_id))
        if can_sin(item.letter):
            parts.append(self.inventory.decode_sin(sin_id))
        return "".join(parts)

    def diacritize(self, text: str) -> str:
        items = decompose(clean_text(text))
        if not items:
            return ""
        if not self.card.pad_to_window:
            return self._diacritize_stitched(items)
        return self._diacritize_v1(items)

    def _diacritize_stitched(self, items: list[HebrewItem]) -> str:
        """v2 inference: overlapping windows, each position predicted by the
        window where it is most central. Windows are char-aligned (no word
        snapping) — attention context makes boundary words a non-issue, and the
        ownership region never touches a window edge."""
        card = self.card
        assert 0 <= card.overlap < card.window, "model card: need 0 <= overlap < window"
        ids = self.inventory.encode_letters("".join(item.normalized for item in items))
        n = len(ids)
        starts = _window_starts(n, card.window, card.overlap)
        # Rows are always full window width (zero-padded): keeps the input shape
        # fixed so trace-exported models with a baked sequence length work too.
        width = card.window
        input_array = np.zeros((len(starts), width), dtype=np.int32)
        for row, s in enumerate(starts):
            input_array[row, : min(width, n - s)] = ids[s : s + width]

        niqqud_ids, dagesh_ids, sin_ids = self._run(input_array)

        cuts = _ownership_cuts(starts, card.window, n)
        out: list[str] = []
        own_start = 0
        for row, s in enumerate(starts):
            for pos in range(own_start, cuts[row]):
                local = pos - s
                out.append(self._emit(items[pos], niqqud_ids[row, local], dagesh_ids[row, local], sin_ids[row, local]))
            own_start = cuts[row]
        return "".join(out).replace(RAFE, "")

    def _diacritize_v1(self, items: list[HebrewItem]) -> str:
        chunks = list(_split_by_length(items, self.card.window))
        if not chunks:
            return ""

        rows = [self.inventory.encode_letters("".join(item.normalized for item in chunk)) for chunk in chunks]
        width = self.card.window
        input_array = np.zeros((len(chunks), width), dtype=np.int32)
        for i, row in enumerate(rows):
            input_array[i, : len(row)] = row

        niqqud_ids, dagesh_ids, sin_ids = self._run(input_array)

        chunk_texts = []
        for chunk_index, chunk in enumerate(chunks):
            chunk_texts.append(
                "".join(
                    self._emit(item, niqqud_ids[chunk_index, pos], dagesh_ids[chunk_index, pos], sin_ids[chunk_index, pos])
                    for pos, item in enumerate(chunk)
                )
            )

        return " ".join(chunk_texts).replace("  ", " ").replace(RAFE, "")

File: packages/nakdimon-js/src/runtime.ts
// Thin ONNX wrapper around the text layer. Deliberately has no hard
// dependency on onnxruntime-web/onnxruntime-node: callers inject a session
// and a Tensor constructor that satisfy the structural types below, so this
// file (and `tsc`) works offline and in either a browser or Node host.
import { RAFE, canDagesh, canNiqqud, canSin, cleanText, decompose, type HebrewItem } from "./hebrew.js";
import { Inventory, type InventoryName } from "./encode.js";

export interface OrtTensorLike {
  data: ArrayLike<number>;
  dims: number[];
}

export interface OrtSessionLike {
  inputNames: string[];
  run(feeds: Record<string, unknown>): Promise<Record<string, OrtTensorLike>>;
}

export type TensorCtor = new (type: string, data: Int32Array, dims: number[]) => unknown;

export interface ModelCard {
  inventory?: InventoryName;
  window?: number;
  padToWindow?: boolean;
  overlap?: number;
}

const DEFAULT_CARD: Required<ModelCard> = {
  inventory: "v1",
  window: 10000,
  padToWindow: true,
  overlap: 0,
};

/** Start offsets of overlapping windows covering [0, n); last window right-aligned. */
export function windowStarts(n: number, window: number, overlap: number): number[] {
  if (n <= window) return [0];
  const stride = window - overlap;
  const starts: number[] = [];
  for (let s = 0; s < n - window; s += stride) starts.push(s);
  starts.push(n - window);
  return starts;
}

/**
 * Exclusive end of each window's owned region: consecutive windows meet at the
 * integer floor midpoint of their overlap, so every position is predicted by
 * the window in which it is most central. Keep identical to the Python port.
 */
export function ownershipCuts(starts: number[], window: number, n: number): number[] {
  const cuts = starts.slice(0, -1).map((s, k) => Math.floor((starts[k + 1] + s + window) / 2));
  cuts.push(n);
  return cuts;
}

/**
 * Ports nakdimon/hebrew.py's split_by_length exactly: greedily fills `out`
 * up to maxlen-1 items, splitting at the last space seen so far rather than
 * mid-word. If no space was seen before maxlen-1 is reached, the whole
 * (word-exceeding-maxlen) run is yielded as-is.
 *
 * `space` is intentionally NEVER reset to the `maxlen` sentinel after a
 * yield — only `is_space` updates it. This means it is often *stale*
 * (an index into a `out` that has since been sliced away) for one or more
 * chunks following a split that wasn't triggered by a space. That
 * staleness is load-bearing: reproducing it exactly (by porting the loop
 * structure verbatim rather than "fixing" it) is required to match v1's
 * chunk boundaries byte-for-byte.
 */
export function splitByLength<T extends { letter: string }>(items: readonly T[], maxlen: number): T[][] {
  if (maxlen <= 1) throw new Error("maxlen must be > 1");
  const chunks: T[][] = [];
  let out: T[] = [];
  let space = maxlen;
  for (const c of items) {
    if (c.letter === " ") space = out.length;
    out.push(c);
    if (out.length === maxlen - 1) {
      chunks.push(out.slice(0, space + 1));
      out = out.slice(space + 1);
    }
  }
  if (out.length > 0) chunks.push(out);
  return chunks;
}

/** argmax over the last axis of a row-major tensor; returns one winning class index per leading-axis "row". */
function argmaxLastAxis(t: OrtTensorLike): Int32Array {
  const classes = t.dims[t.dims.length - 1];
  const rows = t.data.length / classes;
  const out = new Int32Array(rows);
  for (let r = 0; r < rows; r++) {
    const base = r * classes;
    let best = 0;
    let bestVal = -Infinity;
    for (let c = 0; c < classes; c++) {
      const v = t.data[base + c];
      if (v > bestVal) {
        bestVal = v;
        best = c;
      }
    }
    out[r] = best;
  }
  return out;
}

export class Diacritizer {
  private readonly session: OrtSessionLike;
  private readonly Tensor: TensorCtor;
  private readonly inventory: Inventory;
  private readonly window: number;
  private readonly padToWindow: boolean;
  private readonly overlap: number;

  constructor(session: OrtSessionLike, Tensor: TensorCtor, card: ModelCard = {}) {
    const merged = { ...DEFAULT_CARD, ...card };
    this.session = session;
    this.Tensor = Tensor;
    this.inventory = new Inventory(merged.inventory);
    this.window = merged.window;
    this.padToWindow = merged.padToWindow;
    this.overlap = merged.overlap;
  }

  private async runArgmax(inputData: Int32Array, rows: number, width: number) {
    const tensor = new this.Tensor("int32", inputData, [rows, width]);
    const feeds: Record<string, unknown> = { [this.session.inputNames[0]]: tensor };
    const results = await this.session.run(feeds);
    // The model's three output heads are returned in insertion order,
    // matching the ONNX graph's declared output order: niqqud, dagesh, sin.
    const [nOut, dOut, sOut] = Object.values(results);
    return { niqqudIds: argmaxLastAxis(nOut), dageshIds: argmaxLastAxis(dOut), sinIds: argmaxLastAxis(sOut) };
  }

  private emit(item: HebrewItem, niqqudId: number, dageshId: number, sinId: number): string {
    // NFC canonical emission order: niqqud, dagesh, sin-dot (spec mark_order).
    let s = item.letter;
    if (canNiqqud(item.letter)) s += this.inventory.decodeNiqqud(niqqudId);
    if (canDagesh(item.letter)) s += this.inventory.decodeDagesh(dageshId);
    if (canSin(item.letter)) s += this.inventory.decodeSin(sinId);
    return s;
  }

  async diacritize(text: string): Promise<string> {
    const items = decompose(cleanText(text));
    if (items.length === 0) return "";
    if (!this.padToWindow) return this.diacritizeStitched(items);
    return this.diacritizeV1(items);
  }

  /**
   * v2 inference: overlapping char-aligned windows, each position predicted by
   * the window where it is most central (see Python runtime._diacritize_stitched).
   */
  private async diacritizeStitched(items: HebrewItem[]): Promise<string> {
    if (!(this.overlap >= 0 && this.overlap < this.window)) throw new Error("model card: need 0 <= overlap < window");
    const ids = this.inventory.encodeLetters(items.map((it) => it.normalized).join(""));
    const n = ids.length;
    const starts = windowStarts(n, this.window, this.overlap);
    // Rows are always full window width (zero-padded): keeps the input shape
    // fixed so trace-exported models with a baked sequence length work too.
    const width = this.window;
    const inputData = new Int32Array(starts.length * width);
    starts.forEach((s, row) => {
      const len = Math.min(width, n - s);
      for (let i = 0; i < len; i++) inputData[row * width + i] = ids[s + i];
    });

    const { niqqudIds, dageshIds, sinIds } = await this.runArgmax(inputData, starts.length, width);

    const cuts = ownershipCuts(starts, this.window, n);
    const out: string[] = [];
    let ownStart = 0;
    starts.forEach((s, row) => {
      for (let pos = ownStart; pos < cuts[row]; pos++) {
        const flat = row * width + (pos - s);
        out.push(this.emit(items[pos], niqqudIds[flat], dageshIds[flat], sinIds[flat]));
      }
      ownStart = cuts[row];
    });
    return out.join("").replaceAll(RAFE, "");
  }

  private async diacritizeV1(items: HebrewItem[]): Promise<string> {
    const chunks = splitByLength(items, this.window);

    // v1 pads every chunk to the full window width; the bidirectional pass reads
    // padding, so full-width padding is behaviorally load-bearing.
    const padLen = this.window;
    const nChunks = chunks.length;
    const inputData = new Int32Array(nChunks * padLen);
    chunks.forEach((chunk, row) => {
      const ids = this.inventory.encodeLetters(chunk.map((it) => it.normalized).join(""));
      const base = row * padLen;
      ids.forEach((id, col) => {
        inputData[base + col] = id;
      });
      // remaining columns stay 0 (mask/padding token) via Int32Array's zero-init
    });

    const { niqqudIds, dageshIds, sinIds } = await this.runArgmax(inputData, nChunks, padLen);

    const perChunk: string[] = chunks.map((chunk, row) => {
      const base = row * padLen;
      const parts: string[] = [];
      chunk.forEach((item, col) => {
        parts.push(this.emit(item, niqqudIds[base + col], dageshIds[base + col], sinIds[base + col]));
      });
      return parts.join("");
    });

    // v1's predict() does " ".join(chunks) then a single left-to-right pass
    // collapsing every doubled space introduced at chunk boundaries.
    // Python's str.replace already replaces all non-overlapping
    // occurrences in one pass; String.prototype.replaceAll is the exact JS
    // equivalent (not a loop to a fixed point).
    return perChunk.join(" ").replaceAll("  ", " ").replaceAll(RAFE, "");
  }
}

export interface OrtModule {
  InferenceSession: { create(pathOrUrl: string): Promise<OrtSessionLike> };
  Tensor: TensorCtor;
}

/** Convenience wiring for an injected onnxruntime-web (or -node) module. */
export async function createDiacritizer(ort: OrtModule, modelUrl: string, card?: ModelCard): Promise<Diacritizer> {
  const session = await ort.InferenceSession.create(modelUrl);
  return new Diacritizer(session, ort.Tensor, card);
}
